File: ctdb/libctdb/logging.c
/*
   logging wrapper for libctdb

   Copyright (C) Ronnie Sahlberg 2010

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see <http://www.gnu.org/licenses/>.
*/
#include <sys/time.h>
#include <sys/socket.h>
#include <stdio.h>
#include <errno.h>
#include <ctdb.h>
#include <string.h>
#include <tdb.h>
#include "libctdb_private.h"

int ctdb_log_level = LOG_WARNING;

void ctdb_do_debug(struct ctdb_connection *ctdb,
		   int severity, const char *format, ...)
{
        va_list ap;

        va_start(ap, format);
	ctdb->log(ctdb->log_priv, severity, format, ap);
        va_end(ap);
}

/* Attach tdb logging to our ctdb logging. */
void ctdb_tdb_log_bridge(struct tdb_context *tdb,
			 enum tdb_debug_level level,
			 const char *format, ...)
{
	va_list ap;
	int sev;
	struct ctdb_connection *ctdb = tdb_get_logging_private(tdb);
	char *newformat;

	switch (level) {
	case TDB_DEBUG_FATAL:
		sev = LOG_CRIT;
		break;
	case TDB_DEBUG_ERROR:
		sev = LOG_ERR;
		break;
	case TDB_DEBUG_WARNING:
		sev = LOG_WARNING;
		break;
	case TDB_DEBUG_TRACE:
		sev = LOG_DEBUG;
		break;
	default:
		sev = LOG_CRIT;
	}

	if (sev > ctdb_log_level) {
		return;
	}

	newformat = malloc(sizeof("TDB error: ") + strlen(format));
	if (!newformat) {
		DEBUG(ctdb, LOG_ERR,
		      "memory allocation failure reporting tdb error %s",
		      format);
		return;
	}

	/* Prepend TDB error: and remove \n */
	strcpy(newformat, "TDB error: ");
	strcat(newformat, format);
	if (newformat[strlen(newformat)-1] == '\n')
		newformat[strlen(newformat)-1] = '\0';

	va_start(ap, format);
	ctdb->log(ctdb->log_priv, sev, newformat, ap);
	va_end(ap);
	free(newformat);
}

/* Convenient log helper. */
void ctdb_log_file(FILE *outf, int priority, const char *format, va_list ap)
{
	fprintf(outf, "%s:",
		priority == LOG_EMERG ? "EMERG" :
		priority == LOG_ALERT ? "ALERT" :
		priority == LOG_CRIT ? "CRIT" :
		priority == LOG_ERR ? "ERR" :
		priority == LOG_WARNING ? "WARNING" :
		priority == LOG_NOTICE ? "NOTICE" :
		priority == LOG_INFO ? "INFO" :
		priority == LOG_DEBUG ? "DEBUG" :
		"Unknown Error Level");

	vfprintf(outf, format, ap);
	if (priority == LOG_ERR) {
		fprintf(outf, " (%s)", strerror(errno));
	}
	fprintf(outf, "\n");
}

File: python/samba/gp/util/logging.py
#
# samba-gpupdate enhanced logging
#
# Copyright (C) 2019-2020 BaseALT Ltd.
# Copyright (C) David Mulder <dmulder@samba.org> 2022
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import json
import datetime
import logging
import gettext
import random
import sys

logger = logging.getLogger()
def logger_init(name, log_level):
    logger = logging.getLogger(name)
    logger.addHandler(logging.StreamHandler(sys.stdout))
    logger.setLevel(logging.CRITICAL)
    if log_level == 1:
        logger.setLevel(logging.ERROR)
    elif log_level == 2:
        logger.setLevel(logging.WARNING)
    elif log_level == 3:
        logger.setLevel(logging.INFO)
    elif log_level >= 4:
        logger.setLevel(logging.DEBUG)

class slogm(object):
    '''
    Structured log message class
    '''
    def __init__(self, message, kwargs=None):
        if kwargs is None:
            kwargs = {}
        self.message = message
        self.kwargs = kwargs
        if not isinstance(self.kwargs, dict):
            self.kwargs = { 'val': self.kwargs }

    def __str__(self):
        now = str(datetime.datetime.now().isoformat(sep=' ', timespec='milliseconds'))
        args = dict()
        args.update(self.kwargs)
        result = '{}|{} | {}'.format(now, self.message, args)

        return result

def message_with_code(mtype, message):
    random.seed(message)
    code = random.randint(0, 99999)
    return '[' + mtype + str(code).rjust(5, '0') + ']| ' + \
           gettext.gettext(message)

class log(object):
    @staticmethod
    def info(message, data=None):
        if data is None:
            data = {}
        msg = message_with_code('I', message)
        logger.info(slogm(msg, data))
        return msg

    @staticmethod
    def warning(message, data=None):
        if data is None:
            data = {}
        msg = message_with_code('W', message)
        logger.warning(slogm(msg, data))
        return msg

    @staticmethod
    def warn(message, data=None):
        if data is None:
            data = {}
        return log.warning(message, data)

    @staticmethod
    def error(message, data=None):
        if data is None:
            data = {}
        msg = message_with_code('E', message)
        logger.error(slogm(msg, data))
        return msg

    @staticmethod
    def fatal(message, data=None):
        if data is None:
            data = {}
        msg = message_with_code('F', message)
        logger.fatal(slogm(msg, data))
        return msg

    @staticmethod
    def debug(message, data=None):
        if data is None:
            data = {}
        msg = message_with_code('D', message)
        logger.debug(slogm(msg, data))
        return msg
